File: 0350-intersection-of-two-arrays-ii/0350-intersection-of-two-arrays-ii.cpp
class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        //start--initializations
        vector<int> res = {};
        map<int,int> counts = {};
        
        //Make sure nums1 is always the smaller list
        if(nums1.size() > nums2.size())
           swap(nums1,nums2);
        
        //populate the counts dictionary
        for(int num: nums1){
            counts[num]++;
        }
        
        //create the intersection vector
        for(int num: nums2){
            if(counts[num]>0){
                res.push_back(num);
                counts[num]--;
            }
        }
        return res;
    }
};

File: 0350-intersection-of-two-arrays-ii/0350-intersection-of-two-arrays-ii.py
class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        res = []
        if nums2 < nums1: nums1, nums2 = nums2, nums1
        counts = Counter(nums1)
        for num in nums2:
            if counts[num] > 0:
                res.append(num)
                counts[num] -= 1
        return res